agent/api/container/container.go

```diff
@@ -1231,14 +1201,34 @@ func (c *Container) GetTaskARN() string {
 	return c.TaskARNUnsafe
 }
 
-func (c *Container) SetExecCommandAgentMetadata(metadata ExecCommandAgentMetadata) {
-	c.lock.Lock()
-	defer c.lock.Unlock()
-	c.ExecCommandAgentMetadata = metadata
-}
-
-func (c *Container) GetExecCommandAgentMetadata() ExecCommandAgentMetadata {
+// GetManagedAgentByName retrieves the managed agent with the name specified and a boolean indicating whether an agent
+// was found or not.
+// note: a zero value for ManagedAgent if the name is not known to this container.
+func (c *Container) GetManagedAgentByName(agentName string) (ManagedAgent, bool) {
 	c.lock.RLock()
 	defer c.lock.RUnlock()
-	return c.ExecCommandAgentMetadata
+	for _, ma := range c.ManagedAgentsUnsafe {
+		if ma.Name == agentName {
+			return ma, true
+		}
+	}
+	return ManagedAgent{}, false
+}
+
+// UpdateManagedAgentByName updates the state of the managed agent with the name specified. If the agent is not found,
+// no action is taken.
+func (c *Container) UpdateManagedAgentByName(agentName string, state ManagedAgentState) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	for i, ma := range c.ManagedAgentsUnsafe {
+		if ma.Name == agentName {
+			// It's necessary to clone the whole ManagedAgent struct
```

Comment:
why so?

Reply:
ManagedAgentState is an embedded struct. If I do `c.ManagedAgentsUnsafe[i].ManagedState = state` that only updates the temporary `ManagedState` created when I reference it, but doesn't update the real data. 

Another option could be to make this pointers, but I consider that worse because semantically we are saying the data is mutable. 

Yet another option would be to do it as below, but that is more tedious and error prone.

```golang
c.ManagedAgentsUnsafe[i].ID = ...
c.ManagedAgentsUnsafe[i].Status = ...
c.ManagedAgentsUnsafe[i].Reason = ...
c.ManagedAgentsUnsafe[i].Metadata = ...
```

Reply:
discussed offline, feels like a better approach than using locks